api/catalog/api/controllers/elasticsearch/utils.py

```diff
@@ -0,0 +1,160 @@
+from itertools import accumulate
+from math import ceil
+from typing import List, Optional, Tuple
+
+from django.core.cache import cache
+
+from elasticsearch_dsl import Search
+from elasticsearch_dsl.response import Hit, Response
+
+from catalog.api.models import ContentProvider
+from catalog.api.utils.dead_link_mask import get_query_hash, get_query_mask
+from catalog.api.utils.validate_images import validate_images
+
+
+FILTER_CACHE_TIMEOUT = 30
+DEAD_LINK_RATIO = 1 / 2
+ELASTICSEARCH_MAX_RESULT_WINDOW = 10000
+
+
+def _exclude_filtered_providers(s: Search) -> Search:
+    """
+    Hide data sources from the catalog dynamically. This excludes providers with
+    ``filter_content`` enabled from the search results.
+
+    :param s: the search query to issue to Elasticsearch
+    :return: the modified search query
+    """
+
+    filter_cache_key = "filtered_providers"
+    filtered_providers = cache.get(key=filter_cache_key)
+    if not filtered_providers:
+        filtered_providers = ContentProvider.objects.filter(filter_content=True).values(
+            "provider_identifier"
+        )
+        cache.set(
+            key=filter_cache_key,
+            timeout=FILTER_CACHE_TIMEOUT,
+            value=filtered_providers,
+        )
+    if len(filtered_providers) != 0:
+        to_exclude = [f["provider_identifier"] for f in filtered_providers]
+        s = s.exclude("terms", provider=to_exclude)
+    return s
+
+
+def _paginate_with_dead_link_mask(
+    s: Search, page_size: int, page: int
+) -> Tuple[int, int]:
+    """
+    Given a query, a page and page_size, return the start and end
+    of the slice of results.
+
+    :param s: The elasticsearch Search object
+    :param page_size: How big the page should be.
+    :param page: The page number.
+    :return: Tuple of start and end.
+    """
+    query_hash = get_query_hash(s)
+    query_mask = get_query_mask(query_hash)
+    if not query_mask:
+        start = 0
+        end = ceil(page_size * page / (1 - DEAD_LINK_RATIO))
```

Comment:
I'm confused as to what `DEAD_LINK_RATIO` is being used here for, do you mind elaborating on it? (If this is just inherited code that's been moved around and we don't quite understand what's happening feel free to ignore 😅)

Reply:
It was existing code. I don't know what this variable is or does. We can probably leave it untouched for now.

Reply:
Sounds good!